### PR TITLE
[merged] For 0.10.0: Support completions, code actions, validation etc in parameter files (part 1) 

### PR DESCRIPTION
**Pull request by StephenWeatherford**
First half of #566
#534 will be fixed by this

## Details:
**before:** DeploymentTemplate represents a template file and its parsed structure
**after:** DeploymentDocument is base class for DeploymentTemplate and DeploymentParameters (which represents a parameter file)
 
**before:** PositionContext represents a DeploymentTemplate along with a specific position in the doc, along with information which can be gleaned from it (e.g. completions at that position)
**after:** PositionContext is base class for TemplatePositionContext and ParametersPositionContext
 
###	If there's an associated template file:
- Display the associated template file in the status bar
- Right-click -> Open Template File
- Display an error in params file for missing required parameters (those with a default value in the template)
- Show quick fix for "Add missing required parameters"
- Pressing " or CTRL+SPACE triggers completions
- Completion intellisense for adding missing parameters (both required and optional)
- Completion intellisense for adding new parameters that don't existing in the template file
- Go to definition on a parameter takes you to its definition in the template
### If there's no associated template file:
- Completion intellisense for adding new parameters
- Displays "No template file" in status bar
- Comma automatically added before and after the new parameter, as needed

See #566 for more info

### CODE REVIEW

File: assets/armsnippets.jsonc [hunk 1]
```diff
@@ -78,7 +78,7 @@
         "description": "ARM Template Variable"
     },
     "Parameter": {
-        "prefix": "arm-param-value",
+        "prefix": "arm-param",
```
Comment (StephenWeatherford) [inline assets/armsnippets.jsonc]: Made this more consistent

File: package.json [hunk 2]
```diff
@@ -157,16 +160,23 @@
         ],
         "commands": [
             {
+                "$comment": "============= General commands =============",
                 "category": "Azure Resource Manager Tools",
                 "title": "Remove Local Dotnet Core Installation",
                 "command": "azurerm-vscode-tools.uninstallDotnet"
             },
+            {
+                "category": "Azure Resource Manager Tools",
```
Comment (StephenWeatherford) [inline package.json]: This is mostly for the test team

File: src/Configuration.ts [hunk 3]
```diff
@@ -0,0 +1,32 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
+import { ConfigurationTarget, workspace } from "vscode";
```
Comment (StephenWeatherford) [inline src/Configuration.ts]: Added so it can be mocked for unit tests

File: src/DeploymentDocument.ts [hunk 4]
```diff
@@ -0,0 +1,159 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
+import { CodeAction, CodeActionContext, Command, Range, Selection, Uri } from "vscode";
```
Comment (StephenWeatherford) [inline src/DeploymentDocument.ts]: Most of this file comes straight from the original DeploymentTemplate.ts

File: src/JSON.ts [hunk 5]
```diff
@@ -34,20 +35,28 @@ export enum ValueKind {
  * The different types of tokens that can be parsed from a JSON string.
```
Comment (StephenWeatherford) [inline src/JSON.ts]: Functionality changes in this file are mainly:
1) add strict and "enclosed" behavior for finding tokens - the current behavior is wonky ("extended" - if the position is right after a token, it still counts as being on the token) which was intended for completion behavior.  I needed behavior to help determine where parameters can be added based on what object the position is "inside"
2) better functionality for handing comments

File: src/TemplatePositionContext.ts [hunk 6]
```diff
@@ -0,0 +1,657 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
```
Comment (StephenWeatherford) [inline src/TemplatePositionContext.ts]: Most of this file comes from the original PositionContext.ts file.

File: src/parameterFiles/DeploymentFileMapping.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
+
+import * as path from 'path';
```
Comment (StephenWeatherford) [inline src/parameterFiles/DeploymentFileMapping.ts]: Refactored out of parameterFiles.ts

File: src/parameterFiles/DeploymentParameters.ts [hunk 8]
```diff
@@ -0,0 +1,295 @@
+// ----------------------------------------------------------------------------
+// Copyright (c) Microsoft Corporation.  All rights reserved.
+// ----------------------------------------------------------------------------
```
Comment (StephenWeatherford) [inline src/parameterFiles/DeploymentParameters.ts]: New file

File: src/supported.ts [hunk 9]
```diff
@@ -3,12 +3,28 @@
...
+];
+
+export const parameterDocumentSelector = [
+    { language: 'json', scheme: 'file' },
+    { language: 'json', scheme: 'untitled' },
+    { language: 'jsonc', scheme: 'file' },
+    { language: 'jsonc', scheme: 'untitled' },
+];
+
+export const templateOrParameterDocumentSelector = [
```
Comment (StephenWeatherford) [inline src/supported.ts]: Most of AzureTools.ts now registers intellisense providers that work for both JSON files (which may be parameter files) and arm-template files.

**Comment by StephenWeatherford** [comment]
@neilpeterson Please play around with this build.  See the notes at the beginning, including issues.

@alexgav This is ready for review (for 0.10.0).  Sorry it's so big, let me know if you have questions.

File: assets/armsnippets.jsonc (hunk 1 above)
Comment (neilpeterson) [reply assets/armsnippets.jsonc]: Consider also updating here?

https://github.com/microsoft/vscode-azurearmtools/blob/master/assets/jsonsnippets.jsonc#L74

Comment (StephenWeatherford) [reply assets/armsnippets.jsonc]: shouldn't said I fixed it to match the description.

in template:
    "Parameter": {
        "prefix": "**arm-param**",
        "body": [
            "\"${1:parameter1}\": {",
            "   \"type\": \"${2|string,securestring,int,bool,object,secureobject,array|}\",",
            "   \"metadata\": {",
            "        \"description\": \"${3:description}\"",
            "    }",
            "}"
        ],
        "description": "**ARM Template Parameter**"
    },

In params:
    "Parameter Value": {
        "prefix": "**arm-param-value**",
        "body": [
            "\"${1:parameter1}\": {",
            "    \"value\": \"${2:value}\"",
            "}"
        ],
        "description": "**Parameter Value**"
    }

**Pull request merged**